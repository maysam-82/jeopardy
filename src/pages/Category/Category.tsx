import React from 'react';
import { connect } from 'react-redux';
import { Spinner } from 'react-bootstrap';
import { IStoreState } from '../../redux/reducers/index';
import { ICategory } from '../../types/category.d';
import { getClues } from '../../redux/actions/categories';
import history from '../../history';
import Clues from '../../components/Clues';

export interface ICategoryProps {
	selectedCategory: ICategory;
	getClues: Function;
	isFetching: boolean;
}

export class Category extends React.Component<ICategoryProps> {
	componentDidMount() {
		const {
			selectedCategory: { id },
			getClues,
		} = this.props;
		if (id) {
			getClues(id);
		} else {
			history.push('/');
		}
	}
	render() {
		const { selectedCategory, isFetching } = this.props;
		return (
			<div>
				{isFetching ? (
					<div className="spinner">
						<Spinner animation="border" role="status" variant="info" />
					</div>
				) : (
					<Clues title={selectedCategory.title} />
				)}
			</div>
		);
	}
}

const mapStateToProps = (state: IStoreState) => ({
	selectedCategory: state.category.selectedCategory,
	isFetching: state.category.isFetching,
});

export default connect(mapStateToProps, { getClues })(Category);
