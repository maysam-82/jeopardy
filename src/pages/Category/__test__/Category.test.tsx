import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { ICategoryProps, Category } from '../Category';
import { testCategories } from '../../../fixtures/testData';
import { ICategory } from '../../../types/category.d';

const testCategoryProps: ICategoryProps = {
	selectedCategory: testCategories[0],
	isFetching: false,
	getClues: jest.fn(),
};

describe('Category component', () => {
	let component: ShallowWrapper;

	it('should render Category component', () => {
		expect(component).toMatchSnapshot();
	});
	describe('When there is a `selectedCategory`', () => {
		beforeEach(() => {
			component = shallow(<Category {...testCategoryProps} />);
		});
		afterEach(() => {
			jest.clearAllMocks();
		});
		it('should call `getClues` function', () => {
			expect(testCategoryProps.getClues).toHaveBeenCalledTimes(1);
		});
		it('should render `Spinner` while fetching', () => {
			component.setProps({ isFetching: true });
			expect(component.find('Spinner')).not.toBeNull;
		});
		it('should not render `Spinner` after fetching', () => {
			component.setProps({ isFetching: false });
			expect(component.find('Spinner')).toBeNull;
		});
	});
	describe('When there is not a `selectedCategory`', () => {
		const newProps = {
			...testCategoryProps,
			selectedCategory: {} as ICategory,
		};
		component = shallow(<Category {...newProps} />);
		it('should not call `getClues` function', () => {
			expect(newProps.getClues).toHaveBeenCalledTimes(0);
		});
	});
});
