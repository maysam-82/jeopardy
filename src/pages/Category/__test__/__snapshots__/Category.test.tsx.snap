// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category component should render Category component 1`] = `
<div>
  <Connect(Clues) />
</div>
`;
